Update serialization for NSParagraphStyle

sheeparegreat opened the pull request:
#### 6d27ad4515eadd1083bcecda875d94841b977508
<pre>
Update serialization for NSParagraphStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=280772">https://bugs.webkit.org/show_bug.cgi?id=280772</a>
<a href="https://rdar.apple.com/135824523">rdar://135824523</a>

Reviewed by Wenson Hsieh and Sihui Liu.

This change removes the use of native object serialization for NSParagraphStyle

Covered by existing tests:
- Table border color and thickness tested by WKWebView.AttributedStringFromTable
- NSTextTabs tested by AttributedSubstringForProposedRange.TextAlignmentParagraphStyles

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSTextTableSPI.h:
* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::configureNSTextBlockFromParagraphStyleCommonTableAttributes):
(WebCore::reconstructNSTextAlignment):
(WebCore::reconstructNSTextBlockVerticalAlignment):
(WebCore::reconstructNSTextTableLayoutAlgorithm):
(WebCore::reconstructNSWritingDirection):
(WebCore::reconstructStyle):
(WebCore::toNSObject):
(WebCore::extractParagraphStyleAlignment):
(WebCore::extractTextTableBlockVerticalAlignment):
(WebCore::extractTextTableLayoutAlgorithm):
(WebCore::extractParagraphStyleWritingDirection):
(WebCore::extractParagraphStyle):
(WebCore::extractValue):
(WebCore::extractTableBlockAndTableIDs): Deleted.
* Source/WebCore/platform/ios/wak/WAKAppKitStubs.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::shouldEnableStrictMode):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
(TEST(WKWebView, AttributedStringFromTable)):

Canonical link: <a href="https://commits.webkit.org/285172@main">https://commits.webkit.org/285172@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d2dc8bf73a5def15a8b592a74ebe4eea3ebadb6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51096 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24457 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75797 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22888 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73798 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58897 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22708 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56621 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15099 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74749 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46358 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61752 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37070 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43019 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19219 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21229 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64808 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64924 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19582 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77517 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70933 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15917 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18766 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64334 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15961 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61785 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64335 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15859 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12485 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6124 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92718 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46896 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1675 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20442 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47967 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49251 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47709 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,102 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    double paddingMaxXEdge = 0;
+    double paddingMaxYEdge = 0;
+
+    double boarderMinXEdge = 0;
```
border?

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
 struct AttributedStringTextListIDType;
 using AttributedStringTextListID = LegacyNullableObjectIdentifier<AttributedStringTextListIDType>;
 
+enum class TextTableLayoutAlgorithm : uint8_t {
```
bool

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+struct WEBCORE_EXPORT ParagraphStyleTextList {
+    AttributedStringTextListID thisID { };
+    String markerFormat { };
+    long startingItemNumber = 0;
```
There's a recent thread in webkit-dev email list where we have decided to use { 0 } to initialize these instead of = 0

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,202 @@
-        }
+    for (const auto& item : style.textTables) {
+        tables.ensure(item.thisID, [&] {
+            RetainPtr table = adoptNS([(NSTextTable *)[PlatformNSTextTable alloc] init]);
```
cast seems unnecessary.  Would this work?
RetainPtr table = adoptNS([PlatformNSTextTable new])

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,202 @@
+            RetainPtr table = adoptNS([(NSTextTable *)[PlatformNSTextTable alloc] init]);
+            [table setNumberOfColumns:item.numberOfColumns];
+
+            NSTextTableLayoutAlgorithm layout;
```
This should be initialized to a default value if item.layout is an unexpected value.

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,202 @@
+            [table setCollapsesBorders:item.collapsesBorders];
+            [table setHidesEmptyCells:item.hidesEmptyCells];
+
+            const ParagraphStyleCommonTableAttributes& attributes = item;
```
This seems unnecessary.  Just pass item as a parameter in the next line.

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,202 @@
+
+            RetainPtr tableBlock = adoptNS([[PlatformNSTextTableBlock alloc] initWithTable:foundTable->value.get() startingRow:item.startingRow rowSpan:item.rowSpan startingColumn:item.startingColumn columnSpan:item.columnSpan]);
+
+            NSTextBlockVerticalAlignment verticalAlignment;
```
Ditto, initialize this to a default value for if item.verticalAlignment is out of bounds.  Same in a few other places.

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,202 @@
-                adjustedTextBlocks = adoptNS((NSMutableArray<NSTextBlock *> *)[[style textBlocks] mutableCopy]);
-            [adjustedTextBlocks setObject:replacementBlock.get() atIndexedSubscript:index];
+    if (!style.textTableBlockIDs.isEmpty()) {
+        RetainPtr<NSMutableArray> blocks = adoptNS([[NSMutableArray alloc] init]);
```
nit: new is a little more brief than alloc and init.  Same a few other places in this PR.
But in this case, initWithCapacity would be more efficient.  Can createNSArray be used for all of these?
Also, the template parameter of the RetainPtr is redundant and unnecessary.

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -354,26 +481,269 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
+        if (listResult.isNewEntry) {
+            newTextLists.append(ParagraphStyleTextList {
+                listID,
+                [list markerFormat],
```
dot syntax?

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -354,26 +481,269 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
-        return std::optional { std::pair { tableBlockID, tableID } };
-    });
+        if (tableEnsureResults.isNewEntry) {
+            std::optional<RetainPtr<PlatformColor>> backgroundColor;
```
optional RetainPtr seems like it probably has too many states, null being distinct from std::nullopt.

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -354,26 +481,269 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
+                tableID,
+                [nstable numberOfColumns],
+                layout,
+                static_cast<bool>([nstable collapsesBorders]),
```
static_cast is probably unnecessary.  !! is more brief.

achristensen07 commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -354,26 +481,269 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
+    }
+
+    ParagraphStyleWritingDirection writingDirection;
+    switch ([style baseWritingDirection]) {
```
Initialize writingDirection if this has an unexpected value.

szewai commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+
+struct WEBCORE_EXPORT ParagraphStyleTextList {
+    AttributedStringTextListID thisID { };
+    String markerFormat { };
```
I don't think we need default initialize String as it should be default initialized to null string.

szewai commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    std::optional<RetainPtr<PlatformColor>> borderMinXEdgeColor { };
+    std::optional<RetainPtr<PlatformColor>> borderMinYEdgeColor { };
+    std::optional<RetainPtr<PlatformColor>> borderMaxXEdgeColor { };
+    std::optional<RetainPtr<PlatformColor>> borderMaxYEdgeColor { };
```
We don't need to initialize `std::optional`.

szewai commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+struct WEBCORE_EXPORT TextTable : ParagraphStyleCommonTableAttributes {
+    AttributedStringTextTableID thisID { };
+    unsigned long numberOfColumns = 0;
+    TextTableLayoutAlgorithm layout { };
```
We need an explicit value in the initialization of `TextTableLayoutAlgorithm`

szewai commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    long rowSpan = 1;
+    long startingColumn = 0;
+    long columnSpan = 1;
+    TextTableBlockVerticalAlignment verticalAlignment { };
```
We need an explicit value in the initialization of TextTableLayoutAlgorithm

szewai commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    CGFloat marginMinXEdge = 0;
+    CGFloat marginMinYEdge = 0;
+    CGFloat marginMaxXEdge = 0;
+    CGFloat marginMaxYEdge = 0;
```
Should use list-initialization

szewai commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -354,26 +481,269 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
+    case NSTextAlignmentNatural:
+        alignment = WebCore::ParagraphStyleAlignment::Natural;
+        break;
+    }
```
This should be helper function since it's also invoked below.

szewai commented on Source/WebCore/platform/ios/wak/WAKAppKitStubs.h:
```diff
@@ -151,18 +150,21 @@ typedef NSUInteger NSRectEdge;
 #endif
+
+typedef NS_ENUM(NSInteger, NSRectEdge) {
 #ifndef NSMinXEdge
```
Is `ifndef` needed?

sheeparegreat replied:
Doing this keeps the same behavior as previously in the removed code

szewai commented on Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:
```diff
@@ -221,36 +224,48 @@ typedef NS_ENUM(NSUInteger, NSTextTabType) {
     NSDecimalTabStopType
 };
 
+
```
additional blank line?

szewai commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    unsigned long numberOfColumns { 0 };
+    TextTableLayoutAlgorithm layout { TextTableLayoutAlgorithm::Automatic };
+    bool collapsesBorders = false;
+    bool hidesEmptyCells = false;
```
These need to be list initialization too

szewai commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+
+struct WEBCORE_EXPORT TextTab {
+    CGFloat location { 0 };
+    ParagraphStyleAlignment alignment = ParagraphStyleAlignment::Natural;
```
List initialization

szewai commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+struct WEBCORE_EXPORT ParagraphStyle {
+    double defaultTabInterval { 36 };
+    ParagraphStyleAlignment alignment = ParagraphStyleAlignment::Natural;
+    ParagraphStyleWritingDirection writingDirection = ParagraphStyleWritingDirection::Natural;
```
List initialization

szewai commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
```diff
@@ -124,19 +124,121 @@ header: <WebCore/ResourceRequest.h>
+    [Validator='startingRow >= 0'] long startingRow;
+    [Validator='rowSpan >= 0'] long rowSpan;
+    [Validator='startingColumn >= 0'] long startingColumn;
+    [Validator='columnSpan >=0'] long columnSpan;
```
Missing space after ">="

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
```diff
@@ -328,7 +328,14 @@ - (NSAttributedString *)_contentsAsAttributedString
-        EXPECT_TRUE([cell backgroundColor] == expectedBackgroundColor || [[cell backgroundColor] isEqual:expectedBackgroundColor]);
+
+        if (expectedBackgroundColor == nil)
+            EXPECT_EQ([cell backgroundColor], nil);
```
`EXPECT_NULL(cell.backgroundColor);`

szewai commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,191 @@
+    switch (layout) {
+    case TextTableLayoutAlgorithm::Automatic:
+        nsLayout = NSTextTableAutomaticLayoutAlgorithm;
+        break;
```
You could early return without break: `return NSTextTableAutomaticLayoutAlgorithm;`

szewai commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,191 @@
+        nsLayout = NSTextTableAutomaticLayoutAlgorithm;
+        break;
+    case TextTableLayoutAlgorithm::Fixed:
+        nsLayout = NSTextTableFixedLayoutAlgorithm;
```
Early return

szewai commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,191 @@
+        nsLayout = NSTextTableFixedLayoutAlgorithm;
+        break;
+    }
+    return nsLayout;
```
```
ASSERT_NOT_REACHED(); // If we don't expect this case to be reached
return NSTextTableAutomaticLayoutAlgorithm;
```

szewai commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,191 @@
+        nsDirection = NSWritingDirectionNatural;
+        break;
+    }
+    return nsDirection;
```
Same here; early return at each case and then assert not reached for default.

szewai commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -354,26 +470,232 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
+    };
+
+    for (NSTextTab *textTab : [style tabStops]) {
+        newTextTabs.append(TextTab {
```
We could `reserveInitialCapacity` on `newTextTabs` if size is known (i.e. size of tabStops)

szewai commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -354,26 +470,232 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
+
+    for (NSTextTableBlock* item in style.textBlocks) {
+        if (![item isKindOfClass:PlatformNSTextTableBlock])
+            return { };
```
Why do we early return here, but we just skip above (`continue` on line 558)?

szewai commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -41,15 +41,19 @@
-using IdentifierToTableMap = HashMap<AttributedString::TextTableID, NSTextTable *>;
-using IdentifierToTableBlockMap = HashMap<AttributedString::TextTableBlockID, NSTextTableBlock *>;
-using IdentifierToListMap = HashMap<AttributedString::TextListID, NSTextList *>;
+using IdentifierToTableMap = HashMap<AttributedStringTextTableID, RetainPtr<NSTextTable>>;
+using IdentifierToTableBlockMap = HashMap<AttributedStringTextTableBlockID, RetainPtr<NSTextTableBlock>>;
+using IdentifierToListMap = HashMap<AttributedStringTextListID, RetainPtr<NSTextList>>;
```
We could still use `AttributedString::TextTableID`, `AttributedString::TextTableBlockID` and `AttributedString::TextListID`

szewai reviewed (approved):
Looks good to me; we may want @whsieh to double check the `AttributedString` part and ensure the properties are correct.

whsieh commented on Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:
```diff
@@ -227,30 +230,41 @@ typedef NS_ENUM(NSUInteger, NSTextTabType) {
+- (void)setWidth:(CGFloat)val type:(NSTextBlockValueType)type forLayer:(NSTextBlockLayer)layer edge:(NSRectEdge)edge;
+- (CGFloat)valueForDimension:(NSTextBlockDimension)dimension;
+- (CGFloat)widthForLayer:(NSTextBlockLayer)layer edge:(NSRectEdge)edge;
+- (NSTextBlockVerticalAlignment) verticalAlignment;
```
```suggestion
- (NSTextBlockVerticalAlignment)verticalAlignment;
```

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    RetainPtr<PlatformColor> borderMaxYEdgeColor;
+};
+
+struct WEBCORE_EXPORT TextTable : ParagraphStyleCommonTableAttributes {
```
Are these `WEBCORE_EXPORT` necessary? (since the definition seems to be fully contained in the header)

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    bool hidesEmptyCells { false };
+};
+
+struct WEBCORE_EXPORT TextTableBlock : ParagraphStyleCommonTableAttributes {
```
Ditto.

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    TextTableBlockVerticalAlignment verticalAlignment { TextTableBlockVerticalAlignment::Top };
+};
+
+struct WEBCORE_EXPORT TextTab {
```
Ditto

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.h:
```diff
@@ -86,6 +86,108 @@ using AttributedStringTextTableBlockID = LegacyNullableObjectIdentifier<Attribut
+    ParagraphStyleAlignment alignment { ParagraphStyleAlignment::Natural };
+};
+
+struct WEBCORE_EXPORT ParagraphStyle {
```
Ditto

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,177 @@
+        [table setBorderColor:item.borderMaxYEdgeColor.get() forEdge:NSMaxYEdge];
+}
+
+inline static NSTextAlignment reconstructNSTextAlignment(const ParagraphStyleAlignment& alignment)
```
Nit — for a single uint8_t enum class like this, I think we prefer simply passing by value.
```suggestion
inline static NSTextAlignment reconstructNSTextAlignment(ParagraphStyleAlignment alignment)
```

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,177 @@
+    return NSTextAlignmentLeft;
+}
+
+inline static NSTextBlockVerticalAlignment reconstructNSTextBlockVerticalAlignment(const TextTableBlockVerticalAlignment& alignment)
```
Ditto
```suggestion
inline static NSTextBlockVerticalAlignment reconstructNSTextBlockVerticalAlignment(TextTableBlockVerticalAlignment alignment)
```

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,177 @@
+    return NSTextBlockTopAlignment;
+}
+
+inline static NSTextTableLayoutAlgorithm reconstructNSTextTableLayoutAlgorithm(const TextTableLayoutAlgorithm& layout)
```
```suggestion
inline static NSTextTableLayoutAlgorithm reconstructNSTextTableLayoutAlgorithm(TextTableLayoutAlgorithm layout)
```

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,177 @@
+    return NSTextTableAutomaticLayoutAlgorithm;
+}
+
+inline static NSWritingDirection reconstructNSWritingDirection(const ParagraphStyleWritingDirection& writingDirection)
```
```suggestion
inline static NSWritingDirection reconstructNSWritingDirection(ParagraphStyleWritingDirection writingDirection)
```

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,177 @@
+        RetainPtr blocks = createNSArray(style.textTableBlockIDs, [&] (auto& object) -> id {
+            if (tableBlocks.contains(object))
+                return tableBlocks.get(object).get();
+            return nil;
```
There's a redundant hash lookup here — I think you can let this be:
```suggestion
            return tableBlocks.get(object).get();
```

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -85,79 +89,177 @@
+        RetainPtr textLists = createNSArray(style.textListIDs, [&] (auto& object) -> id {
+            if (lists.contains(object))
+                return lists.get(object).get();
+            return nil;
```
Ditto 
```suggestion
            return lists.get(object).get();
```

whsieh commented on Source/WebCore/editing/cocoa/AttributedString.mm:
```diff
@@ -392,13 +687,9 @@ static MultiRepresentationHEICAttachmentData toMultiRepresentationHEICAttachment
     if ([value isKindOfClass:PlatformNSShadow])
         return { { { RetainPtr { (NSShadow *)value } } } };
-    if ([value isKindOfClass:PlatformNSParagraphStyle]) {
+    if ([value isKindOfClass:PlatformNSParagraphStyle])  {
```
Nit - extra whitespace here

```suggestion
    if ([value isKindOfClass:PlatformNSParagraphStyle]) {
```

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
```diff
@@ -328,7 +328,14 @@ - (NSAttributedString *)_contentsAsAttributedString
         EXPECT_EQ([cell widthForLayer:NSTextBlockBorder edge:leftEdge], expectedBorderWidth);
-        EXPECT_TRUE([cell backgroundColor] == expectedBackgroundColor || [[cell backgroundColor] isEqual:expectedBackgroundColor]);
+
+        if (expectedBackgroundColor == nil)
```
Per WebKit style guidelines:

```suggestion
        if (!expectedBackgroundColor)
```

sheeparegreat committed:
Update serialization for NSParagraphStyle
https://bugs.webkit.org/show_bug.cgi?id=280772
rdar://135824523

Reviewed by Wenson Hsieh and Sihui Liu.

This change removes the use of native object serialization for NSParagraphStyle

Covered by existing tests:
- Table border color and thickness tested by WKWebView.AttributedStringFromTable
- NSTextTabs tested by AttributedSubstringForProposedRange.TextAlignmentParagraphStyles

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSTextTableSPI.h:
* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebCore/editing/cocoa/AttributedString.mm:
(WebCore::configureNSTextBlockFromParagraphStyleCommonTableAttributes):
(WebCore::reconstructNSTextAlignment):
(WebCore::reconstructNSTextBlockVerticalAlignment):
(WebCore::reconstructNSTextTableLayoutAlgorithm):
(WebCore::reconstructNSWritingDirection):
(WebCore::reconstructStyle):
(WebCore::toNSObject):
(WebCore::extractParagraphStyleAlignment):
(WebCore::extractTextTableBlockVerticalAlignment):
(WebCore::extractTextTableLayoutAlgorithm):
(WebCore::extractParagraphStyleWritingDirection):
(WebCore::extractParagraphStyle):
(WebCore::extractValue):
(WebCore::extractTableBlockAndTableIDs): Deleted.
* Source/WebCore/platform/ios/wak/WAKAppKitStubs.h:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::shouldEnableStrictMode):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSObject<NSSecureCoding>>):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewGetContents.mm:
(TEST(WKWebView, AttributedStringFromTable)):

Canonical link: https://commits.webkit.org/285172@main
